File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/SemanticModelDiscovery.java
package com.foundgine.core.semantic.metadata;
import com.foundgine.core.semantic.*; import java.util.*;
/** Bridges structural metadata into the provider-neutral semantic model. */
public final class SemanticModelDiscovery {
 private SemanticModelDiscovery(){}
 public static SemanticModel discover(IMetadataCatalog metadata){Objects.requireNonNull(metadata);var entities=new LinkedHashMap<com.foundgine.core.abstractions.EntityId,SemanticEntity>();
  for(var item:metadata.entities()){
   var fields=new ArrayList<SemanticField>();for(var f:item.effectiveFields())fields.add(new SemanticField(f.id(),f.name(),f.clrType(),null,SemanticFieldCapabilities.DEFAULT,f.effectiveAliases().stream().map(a->new SemanticAlias(a.name(),a.weight())).toList(),List.of(),null));
   SemanticField primary=null;if(item.primaryKey()!=null)for(var f:item.effectiveFields())if(f.column()!=null&&f.column().columnId().equals(item.primaryKey().columnId())){primary=fields.stream().filter(x->x.id().equals(f.id())).findFirst().orElse(null);break;}
   if(primary==null)throw new IllegalStateException("Metadata entity '"+item.name()+"' has no field corresponding to its primary key.");
   entities.put(item.entityId(),new SemanticEntity(item.entityId(),item.name(),new SemanticFieldIdentity(primary.id(),primary.name()),fields,List.of(),item.aliases().stream().map(a->new SemanticAlias(a.name(),a.weight())).toList(),item.clrType()));
  }
  for(var r:metadata.relationships()){var s=entities.get(r.source());if(s==null)throw new IllegalStateException("Relationship '"+r.name()+"' references unknown source entity '"+r.source()+"'.");if(!entities.containsKey(r.target()))throw new IllegalStateException("Relationship '"+r.name()+"' references unknown target entity '"+r.target()+"'.");var rs=new ArrayList<>(s.relationships());rs.add(new SemanticRelationship(r.id(),r.name(),r.target(),r.collection()?RelationshipCardinality.MANY:RelationshipCardinality.ONE,r.aliases().stream().map(a->new SemanticAlias(a.name(),a.weight())).toList()));entities.put(r.source(),new SemanticEntity(s.id(),s.name(),s.identity(),s.fields(),rs,s.aliases(),s.modelType()));}
  return new SemanticModel(entities,List.of(),false);
 }
 public static SemanticModelBuilder fromMetadata(IMetadataCatalog metadata){return new SemanticModelBuilder().importModel(discover(metadata));}
}

File: src/csharp/Foundgine.Core/Semantic/Metadata/SemanticModelDiscovery.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>
/// Bridges structural metadata to the semantic model. This lives in
/// Foundgine.Core.Semantic.Metadata (not Foundgine.Core.Semantic) so that the semantic model
/// itself carries no dependency on the concrete metadata assembly; only
/// applications that actually discover semantics from metadata need to
/// reference this type.
/// </summary>
public static class SemanticModelDiscovery
{
    /// <summary>
    /// Discovers the structural semantic model from Foundgine.Core.Semantic.Metadata.
    /// Metadata describes what exists; this method does not grant capability
    /// exposure or authorization. Applications can enrich the result with
    /// logical traversals and policy configuration afterwards.
    /// </summary>
    public static SemanticModel Discover(this IMetadataCatalog metadata)
    {
        ArgumentNullException.ThrowIfNull(metadata);

        var entities = new Dictionary<EntityId, SemanticEntity>();
        foreach (var item in metadata.Entities)
        {
            var fields = item.EffectiveFields
                .Select(field => new SemanticField(
                    field.Id,
                    field.Name,
                    field.ClrType,
                    Aliases: field.Aliases?.Select(a => new SemanticAlias(a.Name, a.Weight)).ToArray()))
                .ToArray();

            var primary = item.PrimaryKey is null
                ? null
                : fields.FirstOrDefault(field =>
                    item.EffectiveFields.Any(source =>
                        source.Id == field.Id &&
                        source.Column?.ColumnId == item.PrimaryKey.ColumnId));

            if (primary is null)
                throw new InvalidOperationException(
                    $"Metadata entity '{item.Name}' has no field corresponding to its primary key.");

            entities[item.EntityId] = new SemanticEntity(
                item.EntityId,
                item.Name,
                new Foundgine.Core.Semantic.SemanticFieldIdentity(primary.Id, primary.Name),
                fields,
                [],
                item.Aliases?.Select(a => new SemanticAlias(a.Name, a.Weight)).ToArray())
            {
                ModelType = item.ClrType
            };
        }

        foreach (var relationship in metadata.Relationships)
        {
            if (!entities.TryGetValue(relationship.Source, out var source))
                throw new InvalidOperationException(
                    $"Relationship '{relationship.Name}' references unknown source entity '{relationship.Source}'.");

            if (!entities.ContainsKey(relationship.Target))
                throw new InvalidOperationException(
                    $"Relationship '{relationship.Name}' references unknown target entity '{relationship.Target}'.");

            var relationships = source.Relationships.ToList();
            relationships.Add(new SemanticRelationship(
                relationship.Id,
                relationship.Name,
                relationship.Target,
                relationship.IsCollection ? RelationshipCardinality.Many : RelationshipCardinality.One,
                relationship.Aliases?.Select(a => new SemanticAlias(a.Name, a.Weight)).ToArray()));

            entities[relationship.Source] = source with
            {
                Relationships = relationships.ToArray()
            };
        }

        return new SemanticModel(entities);
    }

    /// <summary>
    /// Starts a semantic configuration from structural metadata. Ordinary
    /// entities, fields, identities and direct relationships are discovered;
    /// subsequent builder calls are reserved for application meaning.
    /// </summary>
    public static SemanticModelBuilder FromMetadata(this IMetadataCatalog metadata) =>
        new SemanticModelBuilder().Import(metadata.Discover());
}